src/rust/engine/process_execution/src/named_caches.rs

```diff
@@ -0,0 +1,62 @@
+use std::collections::BTreeSet;
+use std::ffi::OsString;
+use std::path::PathBuf;
+
+#[derive(Clone, Debug, Eq, PartialEq, Hash, PartialOrd, Ord)]
+pub struct NamedCache(String);
+
+impl NamedCache {
+  pub fn new(name: String) -> Result<NamedCache, String> {
+    if name.chars().all(|c| c.is_ascii_alphanumeric() || c == '_') {
+      Ok(NamedCache(name.to_lowercase()))
+    } else {
+      Err(format!(
+        "Cache names may only contain alphanumeric characters and underscores: got {:?}",
+        name
+      ))
+    }
+  }
+}
+
+#[derive(Clone)]
+pub struct NamedCaches {
+  ///
+  /// The absolute path to the base of the directory storing named caches. Pants "owns" this
+  /// directory, and may clear or otherwise prune it at any time.
+  ///
+  local_base: PathBuf,
+}
+
+impl NamedCaches {
+  pub fn new(local_base: PathBuf) -> NamedCaches {
+    NamedCaches { local_base }
+  }
+
+  pub fn local_paths<'a>(
+    &'a self,
+    caches: &'a BTreeSet<NamedCache>,
+  ) -> impl Iterator<Item = PathBuf> + 'a {
+    caches
+      .iter()
+      .map(move |cache_name| self.local_base.join(&cache_name.0))
+  }
+
+  ///
+  /// Computes the environment variables that should be set in the environment of a local process
+  /// using the given named caches.
+  ///
+  pub fn local_env<'a>(
```

Comment:
opt: Given `NamedCache` is the thing which internally guarantees that cache names are ASCII, it feels like this struct should maybe be the thing which verifies that `local_base` is valid UTF8 and returns `String`s here, rather than leaving it for the caller to check/convert every time...

Reply:
There are two callsites, and one actually wants things as `OsString`, so I went with exposing them. Can mix that up if the balance changes.